modules/internal/String.chpl

```diff
@@ -1982,19 +2076,13 @@ module String {
       var result: string = this;
       if result.isEmpty() then return result;
 
-      var i = 0;
-      while i < result.buffLen {
-        const (decodeRet, cp, nBytes) = decodeHelp(buff=result.buff,
-                                                   buffLen=result.buffLen,
-                                                   offset=i,
-                                                   allowEsc=false);
+      for (cp, i, nBytes) in this._cpIndexLen() {
         var lowCodepoint = codepoint_toLower(cp);
         if lowCodepoint != cp && qio_nbytes_char(lowCodepoint) == nBytes {
           // Use the MacOS approach everywhere:  only change the case if
```

Comment:
I don't think there is any reason to do this since we are making a new string. We can just count the encoded length of the codepoints to-lower'd and then make a new buffer with that length before running this loop. I think there are a couple of other places this happens. Would you mind fixing that (possibly in another PR?). Anyway with that fix we could skip the count part for an ascii string.

Reply:
Right, what you are saying is basically make this `toLower` et al methods look more like the others like:

```chapel
proc toLower() {
  if this.isASCII() {
    return doToLowerNoEnc(this);  // this helper is also used by `bytes`
  } else {
    // whatever we have on master now
  }
}
```

I prefer to keep it separate from this PR and merge this to start collecting some performance data. But I agree that we should make that change

Reply:
There's more to it - `    // whatever we have on master now` is not what I'm asking for. I'm talking about removing the need for the comment `The case change operation is not currently performed on characters whose cases take different number of bytes to represent in Unicode mapping.` which appears on many of these functions.

Reply:
Anyway it is fine to leave for another PR but let's make an issue for it.

Reply:
Yeah yeah, for sure.. we don't have that workaround for different number of bytes in upper/lower case in the `bytes` version. So, if we have that refactor, we'll have that, too.

Reply:
https://github.com/Cray/chapel-private/issues/1187